contracts/sol/MACI.sol

```diff
@@ -179,8 +210,25 @@ contract MACI is Ownable, DomainObjs {
         PubKey memory _encPubKey
     ) 
     isAfterSignUpDeadline
-    public 
-    {
+    public {
+
+        // When this function is called for the first time, set
+        // postSignUpStateRoot to the last known state root, and destroy the
+        // state tree contract to free up space. We do so as the
+        // batchProcessMessage function can only update the state root as a
+        // variable and has no way to use MerkleTree.insert() anyway.
+        if (postSignUpStateRoot == 0) {
+            // It is exceedingly improbable that the zero value is a tree root
+            assert(postSignUpStateRoot != stateTree.getRoot());
+
+            postSignUpStateRoot = stateTree.getRoot();
+
+            // This is exceedingly unlikely to occur
+            assert(postSignUpStateRoot != 0);
+
+            // Destroy the state tree contract
+            stateTree.selfDestruct();
```

Comment:
I would rather not destory the stateTree because in future we will want to allow signup for ever. But its fine for now.

Reply:
Will remove :smile: